### PR TITLE
Support analyzing cache related spans to provide metrics and slow commands for cache services from client side.

### PR DESCRIPTION
**Pull request by pg-yang**
- [ ] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #<issue number>.
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/docs/en/changes/changes.md).

Related issue  https://github.com/apache/skywalking/issues/9556 。
I‘m not sure whether CICD could pass . In my repo , CICD always need rerun one  time .

### CODE REVIEW

File: docs/menu.yml [hunk 1]
```diff
@@ -111,6 +111,8 @@ catalog:
                 path: "/en/setup/backend/trace-sampling"
               - name: "Detect Slow Database Statement"
                 path: "/en/setup/backend/slow-db-statement"
+              - name: "Detect Slow Database Statement"
```
Comment (wu-sheng) [inline docs/menu.yml:114]: You forgot to update the menu name.

**Review by wu-sheng — COMMENTED** [review]
Are you keeping UI update(for root menu) and dashboard settings for another PR?

File: oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/AnalyzerModuleConfig.java [hunk 2]
```diff
@@ -60,6 +62,23 @@ public class AnalyzerModuleConfig extends ModuleConfig {
...
+    @Getter
+    private CacheWriteLatencyThresholdsAndWatcher cacheWriteLatencyThresholdsAndWatcher;
+
+    @Setter
+    @Getter
+    private String slowCacheReadThreshold = "default:20";
+
+    @Setter
+    @Getter
+    private CacheReadLatencyThresholdsAndWatcher cacheReadLatencyThresholdsAndWatcher;
```
Comment (wu-sheng) [inline oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/AnalyzerModuleConfig.java:80]: With all these new configs and dynamic configuration watchers added, the relative docs should be updated too.

File: oap-server/oal-grammar/src/main/antlr4/org/apache/skywalking/oal/rt/grammar/OALLexer.g4 [hunk 3]
```diff
@@ -45,6 +45,9 @@ SRC_SERVICE_INSTANCE_CLR_GC: 'ServiceInstanceCLRGC';
 SRC_SERVICE_INSTANCE_CLR_THREAD: 'ServiceInstanceCLRThread';
 SRC_ENVOY_INSTANCE_METRIC: 'EnvoyInstanceMetric';
 SRC_EVENT: 'Event';
+SRC_CACHE_READ: 'VirtualCacheRead';
+SRC_CACHE_WRITE: 'VirtualCacheWrite';
```
Comment (wu-sheng) [inline oap-server/oal-grammar/src/main/antlr4/org/apache/skywalking/oal/rt/grammar/OALLexer.g4:49]: Virtual Cache should be one source. And it should include one field named `op`.

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/virtual_cache/virtual-cache-root.json [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            "url": "https://skywalking.apache.org/docs/main/latest/en/setup/service-agent/virtual-cache/"
+          },
+          "moved": false
+        }
+      ],
+      "id": "Virtual-Cache-Root",
+      "layer": "VIRTUAL_CACHE",
+      "entity": "All",
+      "name": "Virtual-Cache-Root",
+      "isRoot": true
```
Comment (wu-sheng) [inline oap-server/server-starter/src/main/resources/ui-initialized-templates/virtual_cache/virtual-cache-root.json:119]: I don't think UI supports this layer as root. Could you provide your local screenshot?

File: test/e2e-v2/java-test-service/e2e-service-provider/src/main/java/org/apache/skywalking/e2e/controller/UserController.java [hunk 5]
```diff
@@ -68,4 +80,39 @@ private long randomSleepLong(int min, int max) {
         int randomNumber = rand.nextInt((max - min) + 1) + min;
         return randomNumber;
     }
+
+    private void testCacheService() {
```
Comment (wu-sheng) [inline test/e2e-v2/java-test-service/e2e-service-provider/src/main/java/org/apache/skywalking/e2e/controller/UserController.java:84]: @kezhenxu94 Do you have recommendation about the test service to include virtual cache testing?

File: test/e2e-v2/script/env [hunk 6]
```diff
@@ -25,4 +25,4 @@ SW_KUBERNETES_COMMIT_SHA=0f3ec68e5a7e1608cec8688716b848ed15e971e5
 SW_ROVER_COMMIT=d956eaede57b62108b78bca48045bd09ba88e653
 SW_BANYANDB_COMMIT=5a326d7e36a008c5ea10e3ae506309cb29733c53
 
-SW_CTL_COMMIT=a7c4ccae106e4a5a0b6296f860396f289444058d
+SW_CTL_COMMIT=fd9b93348f51655ec58046aa645cc127c68bbdd2
```
Comment (wu-sheng) [inline test/e2e-v2/script/env:28]: Why do you revert CLI commit? `fd9b` seems even not existing. What happens?

File: oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/AnalyzerModuleProvider.java [hunk 7]
```diff
@@ -85,15 +90,19 @@ public ModuleConfig createConfigBeanIfAbsent() {
 
     @Override
     public void prepare() throws ServiceNotProvidedException, ModuleStartException {
-        thresholds = new DBLatencyThresholdsAndWatcher(moduleConfig.getSlowDBAccessThreshold(), this);
+        dbLatencyThresholdsAndWatcher = new DBLatencyThresholdsAndWatcher(moduleConfig.getSlowDBAccessThreshold(), this);
+        cacheReadLatencyThresholdsAndWatcher = new CacheReadLatencyThresholdsAndWatcher(moduleConfig.getSlowCacheReadThreshold(), this);
+        cacheWriteLatencyThresholdsAndWatcher = new CacheWriteLatencyThresholdsAndWatcher(moduleConfig.getSlowCacheWriteThreshold(), this);
```
Comment (wu-sheng) [inline oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/AnalyzerModuleProvider.java:95]: Don't put new codes in the first lines. This doesn't have to be here. Please add after `traceSamplingPolicyWatcher`.

File: oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/parser/listener/vservice/VirtualCacheProcessor.java [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        String peer = span.getPeer();
+        // peer is blank if it's a local span
+        if (StringUtil.isBlank(peer)) {
+            peer = tags.get(SpanTags.CACHE_TYPE) + "-local";
+        }
+        long timeBucket = TimeBucket.getMinuteTimeBucket(span.getStartTime());
+        String serviceName = namingControl.formatServiceName(peer);
+        int latency = (int) (span.getEndTime() - span.getStartTime());
+        if (latency > 0) {
+            log.info("latency data : {}", new Gson().toJson(span));
```
Comment (wu-sheng) [inline oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/parser/listener/vservice/VirtualCacheProcessor.java:78]: If you call this, you have to use `log.isInfoEnable` as pre condition, otherwise, `new Gson().toJson(span)` would be called every time, which is very expensive.
Meanwhile, `new Gson` should only be created once as it is a utility.

File: oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/parser/listener/vservice/VirtualCacheProcessor.java [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+            if (latency > config.getCacheWriteLatencyThresholdsAndWatcher().getThreshold(cacheType)) {
+                sourceList.add(parseSlowWrite(segmentObject, latency, serviceName, span, tags));
+            }
+        } else if ("read".equals(op)) {
+            sourceList.add(parseReadMetrics(span, latency, serviceName, timeBucket));
+            if (latency > config.getCacheReadLatencyThresholdsAndWatcher().getThreshold(cacheType)) {
+                sourceList.add(parseSlowRead(segmentObject, latency, serviceName, span, tags));
+            }
+        } else {
+            log.warn("Illegal tag value :{} for 'cache.op' , only write/read would be accepted", op);
```
Comment (wu-sheng) [inline oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/parser/listener/vservice/VirtualCacheProcessor.java:93]: Aren't you having `command` as `op`? And we don't put this warning because there is nothing to warn.

File: oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/parser/listener/vservice/VirtualCacheProcessor.java [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+    private final AnalyzerModuleConfig config;
+
+    private final List<Source> sourceList = new ArrayList<>();
+
+    @Override
+    public void prepareVSIfNecessary(SpanObject span, SegmentObject segmentObject) {
+        if (span.getSpanLayer() != SpanLayer.Cache) {
+            return;
+        }
+        Map<String, String> tags = span.getTagsList().stream()
```
Comment (wu-sheng) [inline oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/parser/listener/vservice/VirtualCacheProcessor.java:63]: I can get the idea why you created `VirtualServiceProcessor`. But there is a side effect, you make every processor implementations iterate span's tags by themselves. This is a performance impact.
Because SkyWalking could process 100b segments, which means 1000-2000b spans per day. The increment of iteration could be a disaster.

File: test/e2e-v2/cases/storage/storage-cases.yaml [hunk 11]
```diff
@@ -47,7 +47,7 @@ cases:
   - query: swctl --display yaml --base-url=http://${oap_host}:${oap_12800}/graphql trace ls
     expected: expected/traces-list.yml
   # negative tags search: relationship should be logical AND instead of logical OR
-  - query: swctl --display yaml --base-url=http://${oap_host}:${oap_12800}/graphql trace ls --tags http.method=POST,http.status_code=200
+  - query: swctl --display yaml --base-url=http://${oap_host}:${oap_12800}/graphql trace ls --tags http.method=POST,http.status_code=200,iserror=true
```
Comment (wu-sheng) [inline test/e2e-v2/cases/storage/storage-cases.yaml:50]: Why do you change this case?

File: test/e2e-v2/cases/meter/meter-cases.yaml [hunk 12]
```diff
@@ -31,4 +31,50 @@
...
+      expected: expected/metrics-has-value0.yml
+    - query: |
+        swctl --display yaml --base-url=http://${oap_host}:${oap_12800}/graphql metrics linear --name=cache_read_sla  --service-id=R3VhdmFDYWNoZS1sb2NhbA==.0  | yq e 'to_entries' -
+      expected: expected/metrics-has-value.yml
+    - query: |
+        swctl --display yaml --base-url=http://${oap_host}:${oap_12800}/graphql metrics linear --name=cache_read_cpm  --service-id=R3VhdmFDYWNoZS1sb2NhbA==.0  | yq e 'to_entries' -
+      expected: expected/metrics-has-value.yml
+    - query: |
+        swctl --display yaml --base-url=http://${oap_host}:${oap_12800}/graphql metrics multiple-linear --name=cache_read_percentile  --service-id=R3VhdmFDYWNoZS1sb2NhbA==.0  | yq e 'to_entries | with(.[] ; .value=(.value | to_entries))' -
+      expected: expected/metrics-has-value-percentile.yml
```
Comment (wu-sheng) [inline test/e2e-v2/cases/meter/meter-cases.yaml:47]: To save cost and time in the CI, I think we don't have to test all metrics of the virtual cache.

You could choose one metric and sampled record.

File: test/e2e-v2/java-test-service/e2e-service-provider/src/main/java/org/apache/skywalking/e2e/controller/UserController.java (hunk 5 above)
Comment (wu-sheng) [reply test/e2e-v2/java-test-service/e2e-service-provider/src/main/java/org/apache/skywalking/e2e/controller/UserController.java:84]: @pg-yang Could you share why do we need so many put/get in this method?

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/DefaultScopeDefine.java [hunk 13]
```diff
@@ -106,6 +107,10 @@ public class DefaultScopeDefine {
     public static final int ZIPKIN_SERVICE_SPAN = 52;
     public static final int ZIPKIN_SERVICE_RELATION = 53;
     public static final int PROCESS_RELATION = 54;
+    public static final int CACHE_READ_ACCESS = 55;
+    public static final int CACHE_WRITE_ACCESS = 56;
+    public static final int CACHE_SLOW_WRITE = 57;
+    public static final int CACHE_SLOW_READ = 58;
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/DefaultScopeDefine.java:113]: These four should be merged into two. The source should include read/write as an attribute rather than using two sources.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/DefaultScopeDefine.java [hunk 14]
```diff
@@ -163,18 +168,18 @@ private static final void addNewScope(ScopeDeclaration declaration, Class origin
         int id = declaration.id();
         if (ID_2_NAME.containsKey(id)) {
             throw new UnexpectedException(
-                "ScopeDeclaration id=" + id + " at " + originalClass.getName() + " has conflict with another named " + ID_2_NAME
-                    .get(id));
+                    "ScopeDeclaration id=" + id + " at " + originalClass.getName() + " has conflict with another named " + ID_2_NAME
+                            .get(id));
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/DefaultScopeDefine.java:172]: All following changes in this file from here, are not necessary. You seem to have different format settings compared with others. Maybe it still could pass format check, but these should not exist in the git changes.

File: docs/en/changes/changes.md [hunk 15]
```diff
@@ -12,6 +12,7 @@
 * Enhance LAL to support converting LogData to DatabaseSlowStatement.
 * [**Breaking Change**] Change the LAL script format(Add layer property).
 * Adapt ElasticSearch 8.1+, migrate from removed APIs to recommended APIs.
+* Add virtual cache to support analyze cache related span
```
Comment (wu-sheng) [inline docs/en/changes/changes.md:15]:
```suggestion
* Support analyzing cache related spans to provide metrics and slow commands for cache services from client side.
```

**Comment by pg-yang** [comment]
Couldn't you visit my comments? My comments are marked `pending`

image

**Comment by wu-sheng** [comment]
You need to submit your comments from this

image

`Pending` means only visible for yourself.

**Review by pg-yang — COMMENTED** [review]
finish my review

File: test/e2e-v2/cases/storage/storage-cases.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (wu-sheng) [reply test/e2e-v2/cases/storage/storage-cases.yaml:50]: What do you mean `negative`?

**Comment by wu-sheng** [comment]
image

You should reply on the same context. Starting a new comment doesn't help reviewer to follow.

**Comment by wu-sheng** [comment]
> The UI code I updated:

Why don't you submit a pull request to update UI right? I can't see a local change would help in any case.

File: test/e2e-v2/cases/meter/meter-cases.yaml (hunk 12 above)
Comment (wu-sheng) [reply test/e2e-v2/cases/meter/meter-cases.yaml:47]: No, that is not testing about. What an e2e test is for checking the workflow of the target service. It costs time and money if you add all cases to verify. 
It is easy to say, let's add all cases, then one day, e2e would end in hours even over a day.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/DefaultScopeDefine.java (hunk 14 above)
Comment (wu-sheng) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/DefaultScopeDefine.java:172]: There is a setting file in the root. Have you imported it?

File: test/e2e-v2/cases/storage/storage-cases.yaml (hunk 11 above)
Comment (wu-sheng) [reply test/e2e-v2/cases/storage/storage-cases.yaml:50]: So? What is wrong? e2e locked the agent version, and the PR you posted has been merged months ago.
e2e is not for every released version, that is not the purpose.

File: oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/parser/listener/vservice/VirtualCacheProcessor.java (hunk 10 above)
Comment (wu-sheng) [reply oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/parser/listener/vservice/VirtualCacheProcessor.java:62]: OK, make sense.

File: oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/parser/listener/vservice/VirtualCacheProcessor.java (hunk 9 above)
Comment (wu-sheng) [reply oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/parser/listener/vservice/VirtualCacheProcessor.java:89]: You should consider a metric call `cache_access_*` to cover all cases(read/write/others). This would show the access payload more accurate.

File: test/e2e-v2/cases/storage/storage-cases.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (pg-yang) [reply test/e2e-v2/cases/storage/storage-cases.yaml:50]: According the  comment in this test case and  the expected data  , `negative ` means query nothing .
The previous java agent  collect the status_code which max  400 , so  the test case could match   expected data .
But the latest java agent collect all http code (include 200), In order to match expected data , I added the param  

You could find this PR https://github.com/apache/skywalking-java/pull/258

BTW , I don't know the meaning of this cache , I'm just keeping this case logically correct

[the remark this answers is not in the document]
Comment (wu-sheng) [reply test/e2e-v2/cases/storage/storage-cases.yaml:50]: statu_code=200, isn't `iserror` should be `false`? 
And `is_error` should be a parameter, rather than a tag.

**Comment by wu-sheng** [comment]
@pg-yang I am going to remove all your duplicated reply. It is really becoming hard to follow the discussions.

**Comment by pg-yang** [comment]
> Why don't you submit a pull request to update UI right? I can't see a local change would help in any case.

Because  I'm not sure I'm allowed to submit PR  to UI , If i'm allowed , I could do it .

>  You should reply on the same context. Starting a new comment doesn't help reviewer to follow.

Very sorry , I added comments to explain why I update  `storage-cases` after submitting  this PR . And it's  marked as `pending ` .
After I  fix `pending ` , I make this PR chaotic . Very sorry .

**Comment by wu-sheng** [comment]
> Because I'm not sure I'm allowed to submit PR to UI , If i'm allowed , I could do it .

There is nothing allowed in an open-source community. If there is only something that may not be correct, then reviewers would guide you out.

File: test/e2e-v2/cases/storage/storage-cases.yaml (hunk 11 above)
Comment (pg-yang) [reply test/e2e-v2/cases/storage/storage-cases.yaml:50]: Before  #258 PR , All http plugins only  collect http code which max 400 , so   the case was meet .
But  since #285 , http plugins collect all  http code , and  I updated `SW_AGENT_JAVA_COMMIT` to latest , so  the the case couldn't meet . In  order keep it correct , I added the param 


BTW  the expected data of this case if  `traces: []` .

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/DefaultScopeDefine.java (hunk 14 above)
Comment (pg-yang) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/DefaultScopeDefine.java:172]: I  don't know `setting file in the root` you said

File: oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/parser/listener/vservice/VirtualCacheProcessor.java (hunk 9 above)
Comment (pg-yang) [reply oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/parser/listener/vservice/VirtualCacheProcessor.java:89]: make sense

File: test/e2e-v2/cases/storage/storage-cases.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (pg-yang) [reply test/e2e-v2/cases/storage/storage-cases.yaml:50]: This case means getting  empty result , I think .
I add the param to keep the result  empty .

Comment (wu-sheng) [reply test/e2e-v2/cases/storage/storage-cases.yaml:50]: Then, why do you use a meaningless tag? Such as `randomKey=randomValue`? `isError` is a segment field, rather than a tag. Using this similar but not the same tag key would make ppl confused.

https://github.com/apache/skywalking-query-protocol/blob/2c541beae657306127f07af127d179492a1d7cf9/trace.graphqls#L28

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/DefaultScopeDefine.java (hunk 14 above)
Comment (wu-sheng) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/DefaultScopeDefine.java:172]: https://github.com/apache/skywalking/blob/master/codeStyle.xml This is for IntelliJ

File: test/e2e-v2/cases/storage/storage-cases.yaml (hunk 11 above)
Comment (pg-yang) [reply test/e2e-v2/cases/storage/storage-cases.yaml:50]: OK, I would choose an appropriate tag

**Comment by pg-yang** [comment]
Could you help me about CICD error  : https://github.com/apache/skywalking/actions/runs/3073171067/jobs/4965295825
I  don't  update the case

**Comment by wu-sheng** [comment]
If you didn't touch relative codes, don't worry. We could recheck/rerun the wrong cases when you have fixed everything. Currently, the CI only matters for yourself to make sure codes are working.

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/virtual_cache/virtual-cache-root.json (hunk 4 above)
Comment (pg-yang) [reply oap-server/server-starter/src/main/resources/ui-initialized-templates/virtual_cache/virtual-cache-root.json:119]: image
image


I would submit PR to UI repo later

Comment (pg-yang) [reply oap-server/server-starter/src/main/resources/ui-initialized-templates/virtual_cache/virtual-cache-root.json:119]: image

image


I  would submit PR to UI repo later

File: test/e2e-v2/java-test-service/e2e-service-provider/src/main/java/org/apache/skywalking/e2e/controller/UserController.java (hunk 5 above)
Comment (pg-yang) [reply test/e2e-v2/java-test-service/e2e-service-provider/src/main/java/org/apache/skywalking/e2e/controller/UserController.java:84]: I have removed some put/get  method .

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/cache/CacheSlowAccessDispatcher.java [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+
+import org.apache.skywalking.oap.server.core.analysis.SourceDispatcher;
+import org.apache.skywalking.oap.server.core.analysis.worker.TopNStreamProcessor;
+import org.apache.skywalking.oap.server.core.source.VirtualCacheSlowAccess;
+
+public class CacheSlowAccessDispatcher implements SourceDispatcher<VirtualCacheSlowAccess> {
+
+    @Override
+    public void dispatch(VirtualCacheSlowAccess source) {
+        if (source.getOp().equals("read")) {
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/cache/CacheSlowAccessDispatcher.java:29]:
```suggestion
        if ("read".equals(source.getOp())) {
```

This is a good practice to avoid NPE.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/cache/CacheSlowAccessDispatcher.java [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+            TopNCacheReadCommand readCommand = new TopNCacheReadCommand();
+            readCommand.setId(source.getId());
+            readCommand.setCommand(source.getCommand() + " " + source.getKey());
+            readCommand.setLatency(source.getLatency());
+            readCommand.setTraceId(source.getTraceId());
+            readCommand.setServiceId(source.getCacheServiceId());
+            readCommand.setTimeBucket(source.getTimeBucket());
+            TopNStreamProcessor.getInstance().in(readCommand);
+        } else {
+            TopNCacheWriteCommand writeCommand = new TopNCacheWriteCommand();
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/cache/CacheSlowAccessDispatcher.java:39]: Do we have other commands? Should we use `else if` here?

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/VirtualCacheAccess.java [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+    private int cacheTypeId;
+    @Getter
+    @Setter
+    private int latency;
+    @Getter
+    @Setter
+    private boolean status;
+    @Getter
+    @Setter
+    private String op; // operation
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/VirtualCacheAccess.java:57]:
```suggestion
    /**
     * op represents an operation of this cache access. 
     */
    @Getter
    @Setter
    private String op;
```

If you want to provide a comment, please provide details and for the field.

Also, I recommend using an enum rather than a string, it has higher performance and is harder to mis-set.  Refer `RequestType` which is in many sources.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/VirtualCacheSlowAccess.java [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+    @Setter
+    private String traceId;
+
+    @Getter
+    @Setter
+    private boolean status;
+
+    @Getter
+    @Setter
+    private String op;
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/VirtualCacheSlowAccess.java:54]: Same recommendation. Please use the enum.

File: oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/CacheReadLatencyThresholdsAndWatcher.java [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+
+    public CacheReadLatencyThresholdsAndWatcher(String config, ModuleProvider provider) {
+        super(AnalyzerModule.NAME, provider, "slowCacheReadThreshold");
+        thresholds = new AtomicReference<>(new HashMap<>());
+        initialSettingsString = config;
+
+        activeSetting(config);
+    }
+
+    private void activeSetting(String config) {
```
Comment (wu-sheng) [inline oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/CacheReadLatencyThresholdsAndWatcher.java:43]: I noticed a potential bug in here, as same as `DBLatencyThresholdsAndWatcher#activeSetting`.

This method doesn't set `default` if it doesn't exist in the provided value. So, in `getThreshold`, a NPE would through as we try to use `thresholds.get().get("default")` for an `int.

We need to fix the existing(also update the change log to add a bug fix), also the new added two watchers.

If there is no default, we should add a default value, as same as we set in the `application.yml`. NOTICE, `initialSettingsString` could be overridden by system env or manual changing. We need to hard code a default value inside this class.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/cache/CacheSlowAccessDispatcher.java (hunk 17 above)
Comment (pg-yang) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/cache/CacheSlowAccessDispatcher.java:39]: We create `VirtualCacheSlowAccess` only  when the op is write/read

Comment (wu-sheng) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/cache/CacheSlowAccessDispatcher.java:39]: OK, but when you change it to an enum, I think doing a `else if` check should be super fast, and make codes better.

File: oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/CacheReadLatencyThresholdsAndWatcher.java (hunk 20 above)
Comment (pg-yang) [reply oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/CacheReadLatencyThresholdsAndWatcher.java:43]: Got it . I would fix these bug together .

File: oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/parser/listener/vservice/VirtualCacheProcessor.java [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+                .collect(Collectors.toMap(KeyStringValuePair::getKey, KeyStringValuePair::getValue));
+        if (!(tags.containsKey(SpanTags.CACHE_KEY) && tags.containsKey(SpanTags.CACHE_OP) && tags.containsKey(SpanTags.CACHE_CMD) && tags.containsKey(SpanTags.CACHE_TYPE))) {
```
Comment (pg-yang) [inline oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/parser/listener/vservice/VirtualCacheProcessor.java:61]: I could change this condition to ` if (!tags.containsKey(SpanTags.CACHE_TYPE))`,
Because if Op ,Command ,key not exist , the `VirtualCacheAccess ` also work .

Comment (wu-sheng) [reply oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/parser/listener/vservice/VirtualCacheProcessor.java:61]: Yes, you could. But we still need to check others to process slow commands when needed.

Also, one nit suggestion, you could use `get(key)` and `not null` instead of `containsKey` to get better performance. `containsKey` + `get` mean you have done `hashcode(key)` and `key match in same hash ID` twice.

File: oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/CacheReadLatencyThresholdsAndWatcher.java (hunk 20 above)
Comment (pg-yang) [reply oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/CacheReadLatencyThresholdsAndWatcher.java:43]: I  have inspect the classes which extends `ConfigChangeWatcher` .  We should update `CacheReadLatencyThresholdsAndWatcher` , `CacheWriteLatencyThresholdsAndWatcher` , `DBLatencyThresholdsAndWatcher`. 
I  think we have two strategy to choose default value  when `get("default")` is Null :

1. Same as `application.yml` 
2. Choose a maximum,  e. g `DBLatencyThresholdsAndWatcher#getThreshold`, we choose `Interger.MAX_VALUE ` as default value  .

Strategy 1 ,  means we use default value from `application.yaml ` if user set default as null .

Strategy 2,  means all condition that depends the  value would be mismatched   (e.g. `if((latency > threshold)` always return false )

Comment (pg-yang) [reply oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/CacheReadLatencyThresholdsAndWatcher.java:43]: I would like Strategy 2 .  Because  If user set property null , it's better that we  could skip  some feature(function) ,
and If user want  some feature(function)  , they should specify appropriate property clearly .

Comment (pg-yang) [reply oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/CacheReadLatencyThresholdsAndWatcher.java:43]: If  we choose Strategy 2 , developers needn't keep the default value consistent  between hard code  and  `application.yaml`.

Comment (wu-sheng) [reply oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/CacheReadLatencyThresholdsAndWatcher.java:43]: Make sense.

**Comment by wu-sheng** [comment]
You seem to have forgotten to update UTs accordingly. UTs fail in compiling.

**Comment by wu-sheng** [comment]
image

According to the screenshots provided in [the UI PR](https://github.com/apache/skywalking-booster-ui/pull/159)
1. Please move access metrics on the top of read and write, and access is more general
2. Have you set an association for metrics among inside access(also read and write). https://skywalking.apache.org/docs/main/next/en/ui/readme/#association-options

**Comment by wu-sheng** [comment]
UI PR merged, please update the submodule and changelog as well.

**Comment by pg-yang** [comment]
> 2\. Have you set an association for metrics among inside access(also read and write). https://skywalking.apache.org/docs/main/next/en/ui/readme/#association-options

I  have set association , but  couldn't take affect .

image

**Comment by wu-sheng** [comment]
You need to click the dot on the line to see association.

**Comment by pg-yang** [comment]
image

image

image

Done

**Comment by wu-sheng** [comment]
Please consider make access on the top, then read on the left(2 lines, 2 metrics per line) and slow SQLs, write on the right?

**Comment by pg-yang** [comment]
image


right?

**Comment by wu-sheng** [comment]
Yes, this should be good.

**Comment by wu-sheng** [comment]
I can't see UI submodule update

**Comment by pg-yang** [comment]
> I can't see UI submodule update
Here

image

**Review by wu-sheng — APPROVED** [review]
You should be able to prepare the showcase update to bring this new feature into the demo env.

**Comment by pg-yang** [comment]
I  have prepared show case ,  after this PR merged , would test on My PC .

File: docs/en/concepts-and-designs/scope-definitions.md [hunk 22]
```diff
@@ -286,3 +286,54 @@ This calculates the metrics data from [events](event.md).
...
+
+This calculates the metrics data from each request of cache system.
+
+| Name | Remarks | Group Key | Type | 
+|---|---|---|---|
+| name | The service name of virtual cache service. |  | string |
+| cacheTypeId | The ID of the component used in this call. | | int |
+| latency | The time taken by each request. | | int(in ms)|
+| status | Indicates the success or failure of the request.| | boolean |
+| op | Indicates this access is used for `write` or `read` | | string |
```
Comment (kezhenxu94) [inline docs/en/concepts-and-designs/scope-definitions.md:324]: This is `operation` in code

```suggestion
| operation | Indicates this access is used for `write` or `read` | | string |
```

File: docs/en/concepts-and-designs/scope-definitions.md [hunk 23]
```diff
@@ -286,3 +286,54 @@ This calculates the metrics data from [events](event.md).
...
+
+| Name | Remarks | Group Key | Type | 
+|---|---|---|---|
+| cacheServiceId | The service id of virtual cache service. |  | string |
+| command | The cache command . | | string |
+| key | The cache command key. | | string|
+| latency | The time taken by each request. | | int(in ms)|
+| traceId | The traceId of this slow access| | string|
+| status | Indicates the success or failure of the request.| | boolean |
+| op | Indicates this access is used for `write` or `read` | | string |
```
Comment (kezhenxu94) [inline docs/en/concepts-and-designs/scope-definitions.md:339]:
```suggestion
| operation | Indicates this access is used for `write` or `read` | | string |
```

**Review by kezhenxu94 — COMMENTED** [review]
@pg-yang can you update the doc in a new PR?

**Comment by pg-yang** [comment]
OK

**Comment by wu-sheng** [comment]
One more thing, the scope should are not as the same as the codes. `VirtualCacheAccess`, `VirtualCacheSlowAccess`
If you want to keep the cache consistent with the database, please remove `virtual` from source names.
